libraries/chain/block_state.cpp

```diff
@@ -193,4 +203,58 @@ std::optional<quorum_certificate> block_state::get_best_qc() const {
    return quorum_certificate{ block_num(), best_qc };
 }
 
+void inject_additional_signatures( signed_block& b, const std::vector<signature_type>& additional_signatures)
+{
+   if (!additional_signatures.empty()) {
+      // as an optimization we don't copy this out into the legitimate extension structure as it serializes
+      // the same way as the vector of signatures
+      static_assert(fc::reflector<additional_block_signatures_extension>::total_member_count == 1);
+      static_assert(std::is_same_v<decltype(additional_block_signatures_extension::signatures), std::vector<signature_type>>);
+
+      emplace_extension(b.block_extensions, additional_block_signatures_extension::extension_id(), fc::raw::pack( additional_signatures ));
+   }
+}
+
+void block_state::sign(const signer_callback_type& signer, const block_signing_authority& valid_block_signing_authority ) {
+   auto sigs = signer( block_id );
+
+   EOS_ASSERT(!sigs.empty(), no_block_signatures, "Signer returned no signatures");
+   block->producer_signature = sigs.back();
+   sigs.pop_back();
+
+   verify_signee(sigs, valid_block_signing_authority);
+   inject_additional_signatures(*block, sigs);
+}
+
+void block_state::verify_signee(const std::vector<signature_type>& additional_signatures, const block_signing_authority& valid_block_signing_authority) const {
+   auto num_keys_in_authority = std::visit([](const auto &a){ return a.keys.size(); }, valid_block_signing_authority);
+   EOS_ASSERT(1 + additional_signatures.size() <= num_keys_in_authority, wrong_signing_key,
+              "number of block signatures (${num_block_signatures}) exceeds number of keys in block signing authority (${num_keys})",
+              ("num_block_signatures", 1 + additional_signatures.size())
+              ("num_keys", num_keys_in_authority)
+              ("authority", valid_block_signing_authority)
+   );
+
+   std::set<public_key_type> keys;
+   keys.emplace(fc::crypto::public_key( block->producer_signature, block_id, true ));
+
+   for (const auto& s: additional_signatures) {
+      auto res = keys.emplace(s, block_id, true);
+      EOS_ASSERT(res.second, wrong_signing_key, "block signed by same key twice", ("key", *res.first));
+   }
+
+   bool is_satisfied = false;
+   size_t relevant_sig_count = 0;
+
+   std::tie(is_satisfied, relevant_sig_count) = producer_authority::keys_satisfy_and_relevant(keys, valid_block_signing_authority);
+
+   EOS_ASSERT(relevant_sig_count == keys.size(), wrong_signing_key,
+              "block signed by unexpected key",
```

Comment:
Missing `${signing_keys)` and `${authority}`. You can also log `relevant_sig_count`